File: theory/data_structures/problems/week3_hash_tables/hash_chains/hashing_with_chains.js

class HashChains {
	constructor() {
		this._prime = 1000000007;
		this._multiplier = 263;
		this.m = 5;
		this._list = [this.m];
		for (let i = 0; i < this.m; i++) {
			this._list[i] = [];
		} 
	}

	hash(entry) {
		let result = 0
		for (let i = 0; i < entry.length; i++) {
			const value = entry.charCodeAt(i);
			result += value * this._multiplier ** i;
		}
		return (result % this._prime) % this.m
	}

	add(entry) {
		let index = this.hash(entry);
		const sublist = this._list[index]
		if (this.hasEntry(sublist, entry)) return
		else sublist.unshift(entry) // add the new entry in the begining of the list
	}

	delete(entry) {
		let index = this.hash(entry);
		const sublist = this._list[index]
		if (this.hasEntry(sublist, entry)) sublist.shift();
	}

	find(entry) {
		let index = this.hash(entry);
		const sublist = this._list[index]
		if (this.hasEntry(sublist, entry)) return "yes";
		else return "no";
	}

	check(index) {
		const sublist = this._list[index];
		let output = "";
		for (let i = 0; i < sublist.length; i++) {
			output += value + " ";
		}
		return output.trim();
	}

	check(index) {
		const sublist = this._list[index];
		let output = "";
		for (let i = 0; i < sublist.length; i++) {
			const value = sublist[i];
			output += value + " ";
		}
		return output.trim();
	}


	hasEntry(sublist, entry) {
		for (let i = 0; i < sublist.length; i++) {
			const value = sublist[i];
			if (entry === value) {
				return true;
			}
		}
		return false;
	}
}

const hc = new HashChains();
hc.add("world")
hc.add("HellO")
console.log(hc.check(4))
console.log(hc.find("World"))
console.log(hc.find("world"))
hc.delete("world")
console.log(hc.check(4))
hc.delete("HellO")
hc.add("luck")
hc.add("GooD")
console.log(hc.check(2))
hc.delete("good")

File: theory/data_structures/problems/week3_hash_tables/hash_chains/hashing_with_chains.py
# python3

class HashChains:
	_prime = 1000000007
	_multiplier = 263
	
	def __init__(self):
		self.m = 5
		self.hashList = []
		for i in range(self.m):
			self.hashList.append([])

	def hash(self, entry):
		result = 0
		for index, value in enumerate(entry):
			result += ord(value) * self._multiplier ** index
		return (result % self._prime) % self.m

	def add(self, entry):
		index = self.hash(entry)
		sublist = self.hashList[index]
		if (self.hasEntry(sublist, entry)):
			return
		else :
			sublist.insert(0, entry)

	def delete(self, entry):
		index = self.hash(entry)
		sublist = self.hashList[index]
		if (self.hasEntry(sublist, entry)):
			sublist.remove(entry)

	def find(self, entry):
		index = self.hash(entry)
		sublist = self.hashList[index]
		if (self.hasEntry(sublist, entry)):
			return "yes"
		else:
			return "no"

	def check(self, index):
		sublist = self.hashList[index]
		output = ""
		for value in sublist:
			output += value + " "
			
		return output.strip()
	
	def hasEntry(self, sublist, entry):
		for value in sublist:
			if entry == value:
				return True
		return False

if __name__ == '__main__':
	hc = HashChains()
	hc.add("world")
	hc.add("HellO")
	print hc.check(4)
	print hc.find("World")
	print hc.find("world")
	hc.delete("world")
	print hc.check(4)
	hc.delete("HellO")
	hc.add("luck")
	hc.add("GooD")
	print hc.check(2)
	hc.delete("good")
	#bucket_count = int(input())
	#proc = QueryProcessor(bucket_count)
	#proc.process_queries()